[clang][analyzer] Add ReportInC99AndEarlier option to DeprecatedOrUnsafeBuf…

gamesh411 opened the pull request:
…ferHandling checker

The checker may report warnings for deprecated buffer handling functions (memcpy, memset, memmove, etc.) even when not compiling with C11 standard if the new option "ReportInC99AndEarlier" is set to true.

These functions are deprecated in C11, but may still be problematic in earlier C standards.

gamesh411 committed:
[clang][analyzer] Add ReportInC99AndEarlier option to DeprecatedOrUnsafeBufferHandling checker

The checker may report warnings for deprecated buffer handling functions
(memcpy, memset, memmove, etc.) even when not compiling with C11 standard
if the new option "ReportInC99AndEarlier" is set to true.

These functions became deprecated in C11, but may still be problematic in
earlier C standards.

gamesh411 committed:
eliminate macro-based test case separation

not quite as suggested, but I think this is expressive now without being
too complicated

gamesh411 committed:
rename flag

gamesh411 committed:
remove redundancy

steakhal commented on clang/lib/StaticAnalyzer/Checkers/CheckSecuritySyntaxOnly.cpp:
```diff
@@ -1113,5 +1116,20 @@ REGISTER_CHECKER(rand)
-REGISTER_CHECKER(DeprecatedOrUnsafeBufferHandling)
+
+void ento::registerDeprecatedOrUnsafeBufferHandling(CheckerManager &mgr) {
+  SecuritySyntaxChecker *checker = mgr.getChecker<SecuritySyntaxChecker>();
```
```suggestion
  SecuritySyntaxChecker *Checker = mgr.getChecker<SecuritySyntaxChecker>();
```

gamesh411 replied:
Fixed

steakhal commented on clang/test/Analysis/security-deprecated-buffer-handling-allow-without-c11.c:
```diff
@@ -0,0 +1,35 @@
+// Test 1: Without C11 and without flag - should NOT warn
+// RUN: %clang_analyze_cc1 %s -verify=c99-noflag -std=gnu99 \
```
So we use `gnu99`; couldn't we use the standard `c99` language flavour instead?
I think the gnu variant allows the gnu extensions, which leads to a strictly broader l language flavour than standard C. Is this intentional?

gamesh411 committed:
rename flag

gamesh411 committed:
fix code style

gamesh411 committed:
Merge branch 'main' into deprecated-buffer-handling-allow-without-c11

gamesh411 committed:
factor out annexk detection

gamesh411 committed:
implement the results of the rfc-discussion

A string option is introduced: ReportMode, which can have 3 values:
"all", "actionable" and "c11-only".

gamesh411 committed:
rename test file

steakhal commented on clang/include/clang/StaticAnalyzer/Checkers/Checkers.td:
```diff
@@ -901,12 +901,25 @@ def UncheckedReturn : Checker<"UncheckedReturn">,
+               "unsafe "
+               "functions regardless of C standard or Annex K availability. "
+               "\"actionable\" only reports when Annex K is available (C11 "
+               "with "
```
I think you should re-flow the comments. Its weird to see a single word in a dedicated line.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/CheckSecuritySyntaxOnly.cpp:
```diff
@@ -754,7 +760,22 @@ void WalkAST::checkDeprecatedOrUnsafeBufferHandling(const CallExpr *CE,
+  const bool ShouldReport = [this] {
+    const bool IsAnnexKAvailable = analysis::isAnnexKAvailable(
+        &BR.getPreprocessor(), BR.getContext().getLangOpts());
+    const bool IsC11OrLaterStandard = BR.getContext().getLangOpts().C11;
```
Do we cache this, or its gonna trigger every time we call `checkDeprecatedOrUnsafeBufferHandling`?
Would that happen only once?

steakhal reviewed (commented):
I like where this is going. Thank you for the detection logic. It looks solid and matches my understanding of Annex K.

gamesh411 committed:
put detection function in a separate cpp file

gamesh411 committed:
reflow checkers.td description

gamesh411 committed:
cache the detection result per TU

gamesh411 committed:
introduce defines to test case for more readability

gamesh411 committed:
Merge branch 'main' into deprecated-buffer-handling-allow-without-c11

NagyDonat reviewed (commented):
Thanks for the commit! At first glance LGTM, I didn't see any obvious issue (but I didn't ponder about details too much).

gamesh411 committed:
fix caching to be per checker and thus per TU effectively

gamesh411 committed:
add StringSwitch

this should does not report with
CheckerManager::reportInvalidCheckerOptionValue, but that one is an
error diagnostic anyway, so emitting an unreachable may not be that much
of a problem.

gamesh411 committed:
refactor test case defines

gamesh411 committed:
fix comment headers

gamesh411 committed:
make request policy validation more robust

gamesh411 commented on clang/test/Analysis/security-deprecated-buffer-handling-report-modes.c:
```diff
@@ -0,0 +1,120 @@
+// REDEFINE: %{std_flag} = %{std_flag_c11}
+// REDEFINE: %{annexk_flag} = %{annexk_defines_unset}
+// REDEFINE: %{config_flag} = %{config_flag_actionable}
+// RUN: %{run_cmd}
```
I think this now is closer to what you meant. I went overboard with the features of lit, this is much more readable now I agree. Thanks for the feedback!

gamesh411 committed:
Merge branch 'main' into deprecated-buffer-handling-allow-without-c11

steakhal reviewed (approved):
This looks great now. I only had a few remarks. One notable would be the handling of invalid checker option - which should not invoke UB.

Given that is fixed, you can merge this.

gamesh411 committed:
reuse test prefix

gamesh411 committed:
Merge branch 'main' into deprecated-buffer-handling-allow-without-c11

gamesh411 committed:
satisfy -Wcovered-switch-default on buildbot
